cudax/examples/stf/linear_algebra/cg_csr.cu

```diff
@@ -115,68 +124,25 @@ public:
   scalar operator-() const
   {
     // Submit a task that computes -s
-    scalar res;
-    ctx.parallel_for(handle.shape(), handle.read(), res.handle.write())
-        ->*[] _CCCL_DEVICE(size_t i, auto dthis, auto dres) {
-              dres(i) = -dthis(i);
-            };
+    scalar res(ctx);
+    res.ctx.parallel_for(box(1), handle.read(), res.handle.write())->*[] _CCCL_DEVICE(size_t i, auto dthis, auto dres) {
+      *dres = -*dthis;
+    };
 
     return res;
   }
 
-  /* This methods reads the content of the logical data in a synchronous manner, and returns its value. */
-  double get() const
-  {
-    double ret;
-    ctx.task(exec_place::host, handle.read())->*[&](cudaStream_t stream, auto dval) {
-      cuda_safe_call(cudaStreamSynchronize(stream));
-      ret = dval(0);
-    };
-    return ret;
-  }
-
-  logical_data<slice<double>> handle;
+  logical_data<scalar_view<double>> handle;
+  context ctx;
 };
 
 scalar DOT(vector& a, vector& b)
 {
-  scalar res;
-
-  /* We initialize the result with a trivial kernel so that we do not need a
-   * cooperative kernel launch for the reduction */
-  ctx.parallel_for(box(1), res.handle.write())->*[] _CCCL_DEVICE(size_t, auto dres) {
-    dres(0) = 0.0;
-  };
-
-  auto spec = par<128>(con<32>());
-  ctx.launch(spec, exec_place::current_device(), a.handle.read(), b.handle.read(), res.handle.rw())
-      ->*[] _CCCL_DEVICE(auto th, auto da, auto db, auto dres) {
-            // Each thread computes the dot product of the elements assigned to it
-            double local_sum = 0.0;
-            for (auto i : th.apply_partition(shape(da), std::tuple<blocked_partition, cyclic_partition>()))
-            {
-              local_sum += da(i) * db(i);
-            }
-
-            auto ti = th.inner();
-            __shared__ double block_sum[th.static_width(1)];
-            block_sum[ti.rank()] = local_sum;
-
-            /* Reduce within blocks */
-            for (size_t s = ti.size() / 2; s > 0; s /= 2)
-            {
-              ti.sync();
-              if (ti.rank() < s)
-              {
-                block_sum[ti.rank()] += block_sum[ti.rank() + s];
-              }
-            }
+  scalar res(a.ctx);
 
-            /* Every first thread of a block writes its contribution */
-            if (ti.rank() == 0)
-            {
-              atomicAdd(&dres(0), block_sum[0]);
-            }
+  a.ctx.parallel_for(a.handle.shape(), a.handle.read(), b.handle.read(), res.handle.reduce(reducer::sum<double>{}))
+      ->*[] __device__(size_t i, auto da, auto db, double& dres) {
+            dres += da(i) * db(i);
```

Comment:
Beautiful simplification! I wonder if we can make the shape, when missing, be the same as the shape of the first argument. In that case this sample would be:

```suggestion
  a.ctx.parallel_for(a.handle.read(), b.handle.read(), res.handle.reduce(reducer::sum<double>{}))
      ->*[] __device__(size_t i, auto da, auto db, double& dres) {
           // shape followed the same as that of `a.handle`, which is the first argument
            dres += da(i) * db(i);
```

Reply:
Did you find that it was already not crowded with overloads ? :D

Reply:
Exhausting overloads are a fact of life in today's C++... see the stdlib. I wonder how clear is for the user.